templated AD NonAD version of the Laromance model with tests.

lynnmunday opened the pull request:
closes #16520   

templated AD and nonAD Laromance models.
For a simple problem with 1000 elements run in serial with a constant stress load that is ramped up for 1e5 seconds of simulation time it shows a 10x speed up over the original ad version.

Commit:
closes #16520  templated AD and nonAd version of the Laromance model with tests.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/649320/) on ebe3c17 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16521/site)

This comment will be updated on new commits.

lynnmunday commented:
@dschwen will you review?

lynnmunday commented:
@tophmatthews would you look.  I switched some of the ADReals to Reals in the AD version of the ADLAROMANCE model.    The ADLAROMANCE model still passes the jacobian tests and other regression tests.   ADLAROMANCE StressUpdateBase::precomputeValues was taking 60% of the runtime when I profiled it.  That function was filling some big ADReal data structures that are grabbing data out of the ROM which I don't think need to be ADReals.  I switched them to Real and the run time goes from being 9x slower than nonAD to 2x slower.  The slowest function now is computeROM and that now only has Real data in it.  The ROM weights are the only data that I think need to be AD.

lynnmunday commented:
@dschwen if the pr "Templated AD/nonAD RadialReturnCreepStressUpdateBase #16471" is merged before this, we can easily template this class too

lynnmunday commented:
@dschwen will you review and merge this pr.  I just updated it.  Then can you help me push the Bison PR that fix all of the bison errors this pr causes?

dschwen commented on modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C:
```diff
@@ -856,21 +892,25 @@ ADLAROMANCEStressUpdateBase::computeTimeStepLimit()
   x = 2.0 * x - 1.0;
-  ADReal plus = std::exp(-2.0 / (1.0 + x));
-  ADReal minus = std::exp(-2.0 / (1.0 - x));
+  GenericReal<is_ad> plus = std::exp(-2.0 / (1.0 + x));
```
you should use more `auto`!!

lynnmunday replied:
> The `auto` comment only affects maybe a dozen lines. This can be fixed later

Can GenericReal<is_ad> be replaced with auto?

dschwen replied:
Yes, as long at the type can be deduced from the right hand side. This wont work if the RHS is a floating point literal, but it will work if the right hand side is an expression containing `GenericReal<is_ad>` typed variables.

dschwen reviewed (approved):
The `auto` comment only affects maybe a dozen lines. This can be fixed later.

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/653524/) on 5c58cca : invalidated by @dschwen

Seems unrelated, let&#39;s see if it turns green.

dschwen commented:
Bison fix is standing by, according to @lynnmunday

tophmatthews commented on modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:
```diff
@@ -64,15 +64,15 @@ class SingleVariableReturnMappingSolution
    * @param scalar                 Inelastic strain increment magnitude being solved for
    */
-  virtual Real computeResidual(const Real effective_trial_stress, const Real scalar) = 0;
+  virtual Real computeResidual(const Real & effective_trial_stress, const Real & scalar) = 0;
```
Aie...this breaks a lot of my uncommitted branches, but it's necessary to template all creep models now, glad to see it in here, even if it is buried in a weird PR...

tophmatthews commented:
I am still skeptical this will work in all situations, or even in situations where things start to really displace, but glad to see some momentum. I'll be interest to follow up on if nonAD works.
